### ISSUE TITLE
Large variants return internal server errors.

### ISSUE THREAD

**Issue by ifokkema** [openvar/variantValidator#151]
Large variants, like this ~200KB duplication, returns an internal server error after a while for both the LOVD and VV endpoints: `NC_000017.10:g.1069645_1279669dup`. If these kinds of variants can not be handled, it would be good to just return an error immediately.

**Comment by PeteCausey-Freeman** [comment]
Linked in to here
https://github.com/openvar/variantValidator/issues/51
Needs to be investigated. Agreed

**Comment by PeteCausey-Freeman** [comment]
Right, I have identified a very inefficient re.search that was taking ages to complete for huge deletions. Vast improvement to time now. There is still a lag, but I will chase it down. 

For reference to me. Variant is 
variant = 'NM_004006.2:c.1483_6438del'
genome_build = 'hg19'

Note the tweak took the time from >20 min to < 1min!!! YAY

**Comment by ifokkema** [comment]
Well done! This kind of optimization is hard to do, but very rewarding when it works :grimacing: 

I must be stressing the test server now, submitting all known DMD variants, many of them very large... :sweat_smile:

**Comment by PeteCausey-Freeman** [comment]
Yeah. I'm gonna fix a few more issues then push this up!

**Comment by ifokkema** [comment]
It seems like we've got quite a performance improvement, thanks!

I cannot thoroughly test yet because the fix for #186 isn't implemented on the server, and I'm not getting hg38 mapping for deletions. This probably speeds up the execution a lot, so I'll recheck when the server gets updated. I do get mappings for large duplications, and there I'm able to submit duplications up to ~225,000 bp (depending on the region, obviously).

**Comment by PeteCausey-Freeman** [comment]
I ran out of time last week. I'll push up the latest version now.

**Comment by PeteCausey-Freeman** [comment]
Ok should be installed. Let me know if not.

If I remember correctly, I also managed to "correct"/improve some additional sub-optimal code last week. So overall for the variant in question I think the improvements have got the validation time down from >15min to <1min. Hey, I'm still learning lol!!! :)

**Comment by ifokkema** [comment]
Haha, well done! Just randomly testing some variant sizes, I'm getting maximum variant sizes of about 250KB duplications or 900KB deletions. What may help is getting rid of these VCF fields as fast in the process as you can, removing a lot of memory overhead, and saving some time on network performance as well. That may require creating a new flag, but it may be worth it! LOVD never uses these fields anyway... :man_shrugging:

**Comment by ifokkema** [comment]
Sorry to nag, but perhaps this one should be reopened; I'm getting some really strange results.
The example you mentioned: `NM_004006.2:c.1483_6438del`, takes [less than a second](https://www35.lamp.le.ac.uk/VariantValidator/variantvalidator/GRCh37/NC_000023.10(NM_004006.2)%3Ac.1483_6438del/all?content-type=application%2Fjson) on the VV endpoint. It's a ~379KB deletion on the genomic level. When trying a smaller, ~252KB deletion in the same gene (`NM_004006.2:c.7309+15980_8218-44279del`), I don't get a reply at all, and [it times out with an internal server error after one minute](https://www35.lamp.le.ac.uk/VariantValidator/variantvalidator/GRCh37/NC_000023.10(NM_004006.2)%3Ac.7309%2B15980_8218-44279del/all?content-type=application%2Fjson). Any clue what's going on? Perhaps the optimizations were really specific?

**Comment by PeteCausey-Freeman** [comment]
No problem,

I'll take a look. In full grant writing mode for the next couple weeks, but after that. Nagging is fine

**Comment by ifokkema** [comment]
For my administration, to later test the fixes; currently failing are:
http://localhost/git/LOVD3/src/scripts/fix_variant_descriptions.php?run&chromosome=X&position=31569850
http://localhost/git/LOVD3/src/scripts/fix_variant_descriptions.php?run&chromosome=X&position=31585083
http://localhost/git/LOVD3/src/scripts/fix_variant_descriptions.php?run&chromosome=X&position=31602163
http://localhost/git/LOVD3/src/scripts/fix_variant_descriptions.php?run&chromosome=X&position=31629880
http://localhost/git/LOVD3/src/scripts/fix_variant_descriptions.php?run&chromosome=X&position=31698023

**Comment by PeteCausey-Freeman** [comment]
Nce idea

**Comment by ifokkema** [comment]
Now that 3.0.0 is out, re-running these tests.

> The example you mentioned: `NM_004006.2:c.1483_6438del`, takes [less than a second](https://rest.variantvalidator.org/VariantValidator/variantvalidator/GRCh37/NC_000023.10(NM_004006.2)%3Ac.1483_6438del/all?content-type=application%2Fjson) on the VV endpoint. It's a ~379KB deletion on the genomic level. When trying a smaller, ~252KB deletion in the same gene (`NM_004006.2:c.7309+15980_8218-44279del`), I don't get a reply at all, and [it times out with an internal server error after one minute](https://rest.variantvalidator.org/VariantValidator/variantvalidator/GRCh37/NC_000023.10(NM_004006.2)%3Ac.7309%2B15980_8218-44279del/all?content-type=application%2Fjson). Any clue what's going on? Perhaps the optimizations were really specific?

🟢 Both variants now work! Response times are 0.8 and 0.7 seconds now, respectively.

🟠 [NC_000023.10:g.31569850_31822113del](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/all/all/False/primary?content-type=application%2Fjson) (took 9 seconds previously, now 37 seconds)
🟢 [NC_000023.10(NM_004006.2):c.7309+15980_8218-44279del](https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NC_000023.10(NM_004006.2)%3Ac.7309%2B15980_8218-44279del/all?content-type=application%2Fjson) (went from a timeout to a 0.7 second response!)

🟠 [NC_000023.10:g.31585085_31803403del](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31585085_31803403del/all/all/False/primary?content-type=application%2Fjson) (took 9 - 25 seconds previously, now 34 seconds)
🟢 [NC_000023.10(NM_004006.2):c.7310-11090_8218-59511del](https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NC_000023.10(NM_004006.2)%3Ac.7310-11090_8218-59511del/all?content-type=application%2Fjson)
(went from a timeout to a 0.7 second response!)

🔴 [NC_000023.10:g.31602164_31980755del](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31602164_31980755del/all/all/False/primary?content-type=application%2Fjson) (took 14 seconds previously, now took 37 seconds and once it failed after 30 seconds)
🟢 [NC_000023.10(NM_004006.2):c.6614+5703_8217+43628del](https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NC_000023.10(NM_004006.2)%3Ac.6614%2B5703_8217%2B43628del/all?content-type=application%2Fjson)  (went from a timeout to a 0.7 second response!)

🟠 [NC_000023.10:g.31698023_32026804del](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31698023_32026804del/all/all/False/primary?content-type=application%2Fjson) (took 12.5 - 26 seconds previously, now 37 seconds)
🟢 [NC_000023.10(NM_004006.2):c.6439-40172_7661-319del](https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NC_000023.10(NM_004006.2)%3Ac.6439-40172_7661-319del/all?content-type=application%2Fjson) (went from a timeout to a 0.7 second response!)

In short, while the VV endpoint got insane speed improvements, the LOVD endpoint is much slower now.

**Comment by ifokkema** [comment]
Hmm... that's an interesting point.
- Ensembl got added, right? So "all" for "transcript_model" used to give me refseq only, but now it gives me refseq+ensembl. That may be a cause.
- On the other hand, back then, "all" for "select_transcripts" worked as "raw" works now, and selecting "all" transcripts could result in fewer transcripts than what "all" used to give me. Anyway, let's look at the data!

**All models, "all" transcripts:**
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/all/all/False/primary?content-type=application%2Fjson
Times out twice after one minute! **Note that, yesterday, this query successfully completed after 37 seconds.**

**Refseq only, "all" transcripts:**
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/all/False/primary?content-type=application%2Fjson
Well, there we go, completes in 6–7 seconds. This is faster than the original 7.5–9 seconds in 2020.

Scaling up a bit again: **refseq only, "raw" transcripts:**
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/raw/False/primary?content-type=application%2Fjson
27 seconds, then 19 seconds on a second run. So "all" is a huge improvement over "raw".

Going down again, refseq only, "select" transcripts:
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/select/False/primary?content-type=application%2Fjson
Strangely enough, 7–10 seconds, while "all" transcripts took 6–7 seconds. There is quite some variability on the server. After checking, I noticed that the results were the same. That's a bit weird, because not all transcripts returned have a select status. Hmm... could you check this? Transcripts that I get back here with the "select" filter that are not select are: NM_000109.4, NM_004009.3, NM_004010.3, NM_004011.4, NM_004012.4, NM_004013.3, NM_004020.4, NM_004021.3, NM_004022.3, NM_004023.3.

**Going down further, refseq only, mane only:**
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/mane/False/primary?content-type=application%2Fjson
Euhm, 7–19.5 seconds. A large variability again, and still no actual filtering of transcripts. The same set as "all" is returned.

**Further testing of the transcript filtering:**
Single nucleotide deletion, refseq only.
- [Raw](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.33038294del/refseq/raw/tx/False?content-type=application%2Fjson): NM_000109.3, NM_000109.4, NM_004006.2, NM_004006.3, NM_004009.3, NM_004010.3.
- [All](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.33038294del/refseq/all/tx/False?content-type=application%2Fjson): NM_000109.4, NM_004006.3, NM_004009.3, NM_004010.3.
- [Select](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.33038294del/refseq/select/tx/False?content-type=application%2Fjson): Identical output as "all".
- [Mane](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.33038294del/refseq/mane/tx/False?content-type=application%2Fjson): Identical output as "all".

**Conclusions**
- Selecting "refseq" instead of "all" transcript models, greatly reduces the required time, and calls can be faster than before.
- However, there is a high variability in response time. Repeated calls can take up to 2.5x longer than the same call takes a few seconds earlier.
- The transcript filtering doesn't seem to work well. "all" shows a large improvement over "raw", but further filtering does not reduce the transcript set.

**Comment by Peter-J-Freeman** [comment]
> Refseq only, "all" transcripts:
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/all/False/primary?content-type=application%2Fjson
> Well, there we go, completes in 6–7 seconds. This is faster than the original 7.5–9 seconds in 2020.

This I would expect. Avoid Raw. It is not really for use outside of searching the entire database for any possible mappings

> Picking a certain transcript:
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/NM_004006/False/primary?content-type=application%2Fjson
8–12 seconds. Strange, not really faster but now the output is restricted... albeit too much, no transcripts are selected at all. 
> Maybe this is what kills the speed improvement; VV needs to handle the variant as intergenic.

Ah, yes, we do need to do something to distinguish between intergenic and not in the select_transcripts. Could really speed this up. Need to open a new issue. 

> Picking a certain transcript + version:
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/NM_004006.3/False/primary?content-type=application%2Fjson
> Now, it's faster; 6–7 seconds. And it selects NM_004006.3 only.

This should be the optimal 

> Going down again, refseq only, "select" transcripts:
https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000023.10%3Ag.31569850_31822113del/refseq/select/False/primary?content-type=application%2Fjson
Strangely enough, 7–10 seconds, while "all" transcripts took 6–7 seconds. There is quite some variability on the server. After checking, I noticed that the results were the same. That's a bit weird, because not all transcripts returned have a select status. Hmm... could you check this? Transcripts that I get back here with the "select" filter that are not select are: NM_000109.4, NM_004009.3, NM_004010.3, NM_004011.4, NM_004012.4, NM_004013.3, NM_004020.4, NM_004021.3, NM_004022.3, NM_004023.3.

I will look again at select filtering. Think it did need a tweak
